llm/memory.go

```diff
@@ -0,0 +1,142 @@
+		opts.NumCtx = max(opts.NumCtx, 2048)
+	}
+
+	// fp16 k,v = (1 (k) + 1 (v)) * sizeof(float16) * n_ctx * n_layer * n_embd / n_head * n_head_kv
+	var kv uint64 = 2 * 2 * uint64(opts.NumCtx) * ggml.KV().BlockCount() * ggml.KV().EmbeddingLength() / ggml.KV().HeadCount() * ggml.KV().HeadCountKV()
+
+	graphPartialOffload, graphFullOffload := ggml.GraphSize(uint64(opts.NumCtx), uint64(min(opts.NumCtx, opts.NumBatch)))
+	if graphPartialOffload == 0 {
+		graphPartialOffload = ggml.KV().GQA() * kv / 6
+	}
+
+	if graphFullOffload == 0 {
+		graphFullOffload = graphPartialOffload
+	}
+
+	graphFullOffload *= uint64(len(gpus))
+	graphPartialOffload *= uint64(len(gpus))
+
+	// memoryRequiredTotal represents the memory required for full GPU offloading (all layers)
+	memoryRequiredTotal := memoryMinimum + graphFullOffload
+
+	// memoryRequiredPartial represents the memory required for partial GPU offloading (n > 0, n < layers)
+	memoryRequiredPartial := memoryMinimum + graphPartialOffload
+
+	if gpus[0].Library != "metal" {
+		if memoryRequiredPartial > memoryAvailable {
+			slog.Debug("insufficient VRAM to load any model layers")
+			return 0, 0, false
+		}
+	}
+
+	var layerCount int
+	layers := ggml.Tensors().Layers()
+	for i := 0; i < int(ggml.KV().BlockCount()); i++ {
+		memoryLayer := layers[fmt.Sprintf("%d", i)].size()
+
+		// KV is proportional to the number of layers
+		memoryLayer += kv / ggml.KV().BlockCount()
+
+		memoryRequiredTotal += memoryLayer
+		if memoryAvailable > memoryRequiredPartial+memoryLayer {
+			memoryRequiredPartial += memoryLayer
+			layerCount++
+		}
+	}
+
+	memoryLayerOutput := layers["output"].size()
+	memoryRequiredTotal += memoryLayerOutput
+
+	systemMemory := uint64(0)
+	if gpus[0].Library == "metal" {
+		memInfo, err := gpu.GetCPUMem()
+		if err != nil {
+			slog.Error("failed to lookup system memory", "error", err)
+		} else {
+			systemMemory = memInfo.TotalMemory
+			slog.Debug("system memory", "totalMB", systemMemory/format.MebiByte)
+		}
+	}
+	if gpus[0].Library == "metal" && memoryRequiredTotal > systemMemory {
+		// disable partial offloading when model is greater than total system memory
+		opts.NumGPU = 0
+	} else if memoryAvailable > memoryRequiredTotal {
+		layerCount = int(ggml.KV().BlockCount()) + 1
+		memoryRequiredPartial = memoryRequiredTotal
+	}
+
+	var allLayers bool
+	if opts.NumGPU < 0 {
+		opts.NumGPU = layerCount
+		allLayers = layerCount >= int(ggml.KV().BlockCount()+1)
+	} else {
+		allLayers = layerCount >= opts.NumGPU
+	}
+
+	slog.Info(
+		"offload to gpu",
+		"reallayers", opts.NumGPU,
+		"layers", layerCount,
+		"required", format.HumanBytes2(memoryRequiredTotal),
+		"used", format.HumanBytes2(memoryRequiredPartial),
+		"available", format.HumanBytes2(memoryAvailable),
+		"kv", format.HumanBytes2(kv),
+		"fulloffload", format.HumanBytes2(graphFullOffload),
+		"partialoffload", format.HumanBytes2(graphPartialOffload),
+	)
+	return layerCount, uint64(memoryRequiredTotal), allLayers
```

Comment:
`memoryRequiredTotal` is likely not what you want. It represents the memory required to offload all layers. Instead, `memoryRequiredPartial` represents the memory required to offload `layerCount` layers. If `layerCount >= opts.NumGPU`, `memoryRequiredPartial` will be the same as `memoryRequiredTotal`